### PR TITLE
[HIP] [OPUS] [JIT] Add GPT-OSS IQ2R 2-bit MoE support

### PR DESCRIPTION
**Pull request by ssharma4-amd**
## Summary

- Add a self-contained AITER implementation of the GPT-OSS O0/IQ2R approximately 2.39-bit expert-weight format, encoder, checkpoint ABI, loader, and overlay support.
- Add gfx950 IQ2R expert GEMMs plus fused top-k routing, stable task construction, dynamic MXFP8 packing, GPT-OSS SwiGLU/packing, and route-weighted BF16 reduction.
- Add graph-safe GPT-OSS TP1/EP1 orchestration through `QuantType.iq2r_2bit`.
- Add captured-production-route correctness and performance harnesses.
- Optimize decode and 16K-scheduler prefill shapes with route-qualified projection-specific dispatch.

## Scope

This is strictly a GPT-OSS MoE change. It does not modify or quantize QKV, O-projection, attention, KV-cache behavior, or LM-head execution. Activations use dynamic per-1x32 MXFP8/E8M0 scaling; only expert weights use IQ2R. The implementation has no Redline source or runtime dependency.

Companion integration: ROCm/ATOM#2295.

## Isolated complete-MoE performance

The timed boundary begins with BF16 hidden states and FP32 router logits and ends after top-k, route/task construction, input packing, both expert GEMMs, SwiGLU, and BF16 route-weighted reduction.

Compared with matched dynamic-MXFP8 Triton A8W4 on MI355X/gfx950:

| Token M | IQ2R speedup | Captures won |
|---:|---:|---:|
| 5 | 1.392x | 72/72 |
| 6 | 1.360x | 72/72 |
| 7 | 1.381x | 72/72 |
| 8 | 1.391x | 72/72 |
| 2048 | 1.221x | 36/36 |
| 4096 | 1.197x | 36/36 |
| 6556 | 1.269x | 12/12 |
| 8000 | 1.226x | 12/12 |
| 13993 | 1.159x | 12/12 |
| 16383 | 1.141x | 12/12 |

All captured route IDs reproduced exactly. The selected decode families are bit-identical to the previous IQ2R family outputs.

## ATOM production translation

Protocol: GPT-OSS-120B, one MI355X/gfx950, TP1/EP1, random 80%-100% of 8192 input and 1024 output tokens, FP8 KV cache, prefix cache disabled, `max_num_batched_tokens=16384`, fresh server per point, and `HIP_VISIBLE_DEVICES=6`.

### C1-C64 versus matched dynamic-MXFP8 Triton A8W4

| C | Dynamic A8W4 TPOT | IQ2R TPOT | TPOT reduction | Output-throughput gain |
|---:|---:|---:|---:|---:|
| 1 | 3.6098 ms | 3.2056 ms | 11.19% | 12.05% |
| 2 | 4.0005 ms | 3.5980 ms | 10.06% | 10.87% |
| 4 | 4.7454 ms | 4.2599 ms | 10.23% | 11.27% |
| 8 | 6.3556 ms | 5.7292 ms | 9.86% | 10.61% |
| 16 | 7.9276 ms | 6.4381 ms | 18.79% | 22.14% |
| 32 | 11.1174 ms | 9.4827 ms | 14.70% | 16.63% |
| 64 | 16.2516 ms | 14.2556 ms | 12.28% | 13.61% |

Geometric mean over C1-C64: **1.1390x TPOT speedup**, **12.21% lower TPOT**, and **13.46% higher output throughput**. C64 uses independent three-run medians.

### C128-C256 versus current stock AITER MXFP4

The development image cannot link the FlyDSL A4W4 prefill kernel at these large serving shapes. The controlled runnable stock baseline therefore uses AITER's two-stage BF16-activation/MXFP4-weight MoE module, while IQ2R uses `module_iq2r_moe.so`.

| C | Stock MXFP4 TPOT | IQ2R TPOT | TPOT reduction | Stock output tok/s | IQ2R output tok/s | Gain |
|---:|---:|---:|---:|---:|---:|---:|
| 128 | 32.1942 ms | 24.2048 ms | **24.82%** | 3,765.78 | 5,026.58 | **33.48%** |
| 256 | 57.3336 ms | 40.5251 ms | **29.32%** | 4,238.03 | 6,035.79 | **42.42%** |

The C128 and C256 runs completed 1,280 and 2,560 requests per path.

## Published InferenceX reference

Against the January 2026 ATOM 0.1.1 FP4 result, final IQ2R is +6.14% in median interactivity and +7.32% in output throughput at C64, and +2.16% / +0.85% at C128. This is a cross-image absolute comparison, not a controlled uplift claim. The published run has no matching TP1/EP1 C256 point.

## Validation and current release status

Final AITER head `8c1c42f08`, run on MI355X/gfx950:

```text
54 focused AITER IQ2R GPU tests passed
14 focused ATOM IQ2R integration tests passed against the companion branch
captured route IDs matched
module_iq2r_moe.so load and execution confirmed
ruff lint and format checks passed
git diff --check passed
```

Sequential real-chat endpoint checks pass 3/3 for both stock and IQ2R. A new exact simultaneous C1-C256 semantic matrix found occasional repetitive/corrupt tails in **both** current stock MXFP4 and IQ2R: stock passed 498/511 after correcting one validator-only false failure, and IQ2R passed 491/511. The issue remains with eager execution, unique prompts, and BF16 KV, while the historical ATOM 0.1.1 stock C8 control is semantically 8/8. This currently points to a shared current-stack concurrent-serving regression rather than an IQ2R-only numerical defect.

Performance acceptance passes. Production coherence signoff remains blocked until that shared serving regression is isolated and the C1-C256 semantic matrix passes cleanly.

The AITER and ATOM worktrees are clean. Temporary benchmark, dense-path, and baseline-only diagnostic files are not part of either PR.

### CODE REVIEW

**Comment by ssharma4-amd** [comment]
Large-route follow-up pushed in d7ac5a98f.

- Replaces the general stable-rank route sorter with an O(routes) shared-memory histogram/prefix/atomic-cursor sorter. Scatter indices preserve original top-k order, so within-expert stability is not required.
- Selects 4x-CU for GPT-OSS down projection at 256+ routed rows.
- Full IQ2R suite on MI355X/gfx950 with HIP_VISIBLE_DEVICES=6: 59 passed.

Matched ATOM serving, GPT-OSS 120B TP1/EP1, FP8 KV, 8192 input / 1024 output, max batched tokens 1024, three repetitions:

| C | dynamic A8W4 | IQ2R | IQ2R TPOT reduction |
|---:|---:|---:|---:|
| 64 | 18.269 ms | 16.537 ms | +9.5% |
| 128 | 29.776 ms | 28.915 ms | +2.9% |
| 256 | 50.369 ms | 51.587 ms | -2.4% |

Versus the previous IQ2R build, TPOT improves 10.1%, 11.8%, and 13.9% at c64/c128/c256. All 36 captured c256 layer routes reproduce exactly; isolated post-router replay is 1.036x median faster than dynamic A8W4 with 34/36 wins. The remaining c256 deficit appears only in full-model graph replay and is not fixed by further launch-family/grid tuning.

**Comment by ssharma4-amd** [comment]
Final qualification update (2026-09-19):

- rebased onto current AITER main; PR is mergeable
- removed the rejected direct down-GEMM/cross-CTA reduction experiment (0.427x-0.531x of control)
- retained parallel8 SwiGLU+MXFP8 and the routed M=20..32 3x4,5xCU down dispatch
- 92 focused IQ2R tests pass on HIP_VISIBLE_DEVICES=6
- post-rebase captured-route smoke: M=5 1.145x, M=8 1.235x versus dynamic A8W4; 2/2 routes won at each size with exact route IDs
- full captured suite remains 1.207x/1.178x/1.345x/1.119x/1.256x/1.324x at M=1/2/4/5/8/16
- ATOM c1-c64 geometric-mean TPOT speedup is 1.1666x versus Redline's 1.1578x O0-over-P2 reference

Head: `546a757574de75b8be1b357fe4b7be2b442702f1`

**Comment by ssharma4-amd** [comment]
Exact-head follow-up: AITER `546a75757` + ATOM `613e5690` completed the graph-mode GPT-OSS 120B smoke twice with identical token IDs and sampled-token log probabilities (`max delta = 0.0`). The pinned local container required disabling the current-main tuned BF16 table because an unrelated FlyDSL dense warmup kernel failed to link; IQ2R remained enabled and loaded from `module_iq2r_moe.so`.

**Comment by ssharma4-amd** [comment]
Final cleanup/qualification update (2026-09-20):

- rebased onto ROCm/aiter `main` at `9f9cc3a41`
- current head: `9024772e8`
- resolved the dependency-pin overlap by keeping upstream `flydsl==0.3.4.1` and the IQ2R `safetensors==0.8.0` requirement
- removed rejected experimental production code; no O1/QTIP additions are present
- stabilized the fused reduction/RMSNorm check at one BF16 ULP
- 92/92 focused IQ2R tests pass on MI355X/gfx950 with `HIP_VISIBLE_DEVICES=6`
- ruff and `git diff --check` pass

The PR description now contains the consolidated full-MoE and ATOM c1-c256 results, provenance, and the explicit pre-rebase serving-data caveat.
